river/tree/_attribute_observer/numeric_attribute_regression_quantizer_observer.py

```diff
@@ -0,0 +1,163 @@
+
+from river.stats import Mean, Var
+from river.utils import VectorDict
+
+from .._attribute_test import AttributeSplitSuggestion, NumericAttributeBinaryTest
+from .attribute_observer import AttributeObserver
+
+
+class NumericAttributeRegressionQuantizerObserver(AttributeObserver):
+    """Quantizer observer (QO).
+
+    Utilizes a dynamical hash-based quantization algorithm to keep track of the target statistics
+    and evaluate split candidates. This class implements the algorithm described in [^1].
+    This attribute observer keeps an internal estimator of the input feature's variance. By doing
+    that, QO can calculate better values for its radius parameter to be passed to future learning
+    nodes.
+
+    Parameters
+    ----------
+    radius
+        The quantization radius.
+
+    References
+    ----------
+    [^1]: Mastelini, S.M. and de Leon Ferreira, A.C.P., 2021. Using dynamical quantization to
+    perform split attempts in online tree regressors. Pattern Recognition Letters.
+
+    """
+
+    def __init__(self, radius: float = 0.01):
+        super().__init__()
+        self.radius = radius if radius > 0 else 0.01
+        self._x_var = Var()
+        self._quantizer = FeatureQuantizer(radius=self.radius)
+
+    def update(self, x, y, sample_weight):
+        if x is None:
+            return
+        else:
+            self._x_var.update(x, sample_weight)
+            self._quantizer.update(x, y, sample_weight)
+
+    def probability_of_attribute_value_given_class(self, x, y):
+        raise NotImplementedError
+
+    def best_evaluated_split_suggestion(
+        self, criterion, pre_split_dist, att_idx, binary_only=True
+    ):
+        candidate = AttributeSplitSuggestion(None, [{}], -math.inf)
+
+        p_x = None
+        for (x, left_dist) in self._quantizer:
+            # First hash element
+            if p_x is None:
+                # In case the hash carries just one element return the null split
+                if len(self._quantizer) == 1:
+                    return candidate
+                p_x = x
+                continue
+
+            right_dist = pre_split_dist - left_dist
+            post_split_dists = [left_dist, right_dist]
+            merit = criterion.merit_of_split(pre_split_dist, post_split_dists)
+
+            if merit > candidate.merit:
+                split_point = (p_x + x) / 2.0
+                candidate = self._update_candidate(
+                    split_point, att_idx, post_split_dists, merit
+                )
+
+            p_x = x
+        return candidate
+
+    @property
+    def x_var(self):
+        return self._x_var
+
+    @staticmethod
+    def _update_candidate(split_point, att_idx, post_split_dists, merit):
+        num_att_binary_test = NumericAttributeBinaryTest(att_idx, split_point, True)
+        candidate = AttributeSplitSuggestion(
+            num_att_binary_test, post_split_dists, merit
+        )
+
+        return candidate
+
+
+class Slot:
+    def __init__(
```

Comment:
Docstring missing. Since it is an internal class just a general description should be enough.

Reply:
Description added in commit [fec06b8](https://github.com/online-ml/river/pull/476/commits/fec06b85a8b38f880bc9286c2ef570edc549717f).